gateway/src/main/scala/filodb/gateway/conversion/InputRecord.scala

```diff
@@ -119,17 +120,14 @@ object PrometheusInputRecord {
    * If a tag in copyTags is found and the destination tag is missing, then the destination tag is created
    * with the value from the source tag.
    */
-  def transformTags(tags: Seq[(String, String)], options: DatasetOptions): Seq[(String, String)] = {
-    val keys = tags.map(_._1).toSet
-    val extraTags = new collection.mutable.ArrayBuffer[(String, String)]()
-    tags.foreach { case (k, v) =>
-      if (options.copyTags contains k) {
-        val renamedKey = options.copyTags(k)
-        if (!(keys contains renamedKey))
-          extraTags += renamedKey -> v
+  def transformTags(tags: collection.mutable.HashMap[String, String],
+                    options: DatasetOptions): collection.mutable.Map[String, String] = {
+    for ((k, v) <- options.copyTags) {
```

Comment:
Here is how you can write an immutable version of transformTags: filter and create the extra tags, then add them to the base collection.  Advantage: we keep the output immutable and free of potential mutation bugs.

    val extraTags = options.copyTags.collect {
      case (k, v) if !tags.contains(v) && tags.contains(k) => v -> tags(k)
    }
    tags ++ extraTags

Reply:
The above code is overriding until all the items in copyTags list is being iterated through. We do not want that to happen since we want to honor the order. Hence, i added the following check : 
```    for ((k, v) <- options.copyTags) {
      **if (!extraTags.contains(v)**
        && !tagsMap.contains(v)
        && tagsMap.contains(k)) {
        extraTags += v -> tagsMap(k)
      }
    }
```